File: Android/src/build-backups/kapt3-20260412-000124/stubs/debug/selfgemma/talk/feature/roleplay/sessions/SessionsViewModel.java
package selfgemma.talk.feature.roleplay.sessions;

@kotlin.Metadata(mv = {2, 2, 0}, k = 1, xi = 48, d1 = {"\u0000b\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0010\u0002\n\u0000\n\u0002\u0010\u000e\n\u0002\b\u0007\n\u0002\u0010\b\n\u0000\n\u0002\u0010\u0011\n\u0002\u0010\u0000\n\u0002\b\u0003\b\u0007\u0018\u0000 \'2\u00020\u0001:\u0001\'B;\b\u0007\u0012\b\b\u0001\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u0012\u0006\u0010\b\u001a\u00020\t\u0012\u0006\u0010\n\u001a\u00020\u000b\u0012\u0006\u0010\f\u001a\u00020\r\u00a2\u0006\u0004\b\u000e\u0010\u000fJ\u000e\u0010\u0017\u001a\u00020\u00182\u0006\u0010\u0019\u001a\u00020\u001aJ\u000e\u0010\u001b\u001a\u00020\u00182\u0006\u0010\u0019\u001a\u00020\u001aJ\u000e\u0010\u001c\u001a\u00020\u00182\u0006\u0010\u0019\u001a\u00020\u001aJ\u0016\u0010\u001d\u001a\u00020\u00182\u0006\u0010\u0019\u001a\u00020\u001a2\u0006\u0010\u001e\u001a\u00020\u001aJ\u0016\u0010\u001f\u001a\u00020\u00182\u0006\u0010\u0019\u001a\u00020\u001a2\u0006\u0010\u001e\u001a\u00020\u001aJ+\u0010 \u001a\u00020\u001a2\b\b\u0001\u0010!\u001a\u00020\"2\u0012\u0010#\u001a\n\u0012\u0006\b\u0001\u0012\u00020%0$\"\u00020%H\u0002\u00a2\u0006\u0002\u0010&R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\n\u001a\u00020\u000bX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\f\u001a\u00020\rX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0014\u0010\u0010\u001a\b\u0012\u0004\u0012\u00020\u00120\u0011X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0017\u0010\u0013\u001a\b\u0012\u0004\u0012\u00020\u00120\u0014\u00a2\u0006\b\n\u0000\u001a\u0004\b\u0015\u0010\u0016\u00a8\u0006("}, d2 = {"Lselfgemma/talk/feature/roleplay/sessions/SessionsViewModel;", "Landroidx/lifecycle/ViewModel;", "appContext", "Landroid/content/Context;", "conversationRepository", "Lselfgemma/talk/domain/roleplay/repository/ConversationRepository;", "roleRepository", "Lselfgemma/talk/domain/roleplay/repository/RoleRepository;", "ensureRoleplaySeedData", "Lselfgemma/talk/domain/roleplay/usecase/EnsureRoleplaySeedDataUseCase;", "importStChatJsonlIntoSessionUseCase", "Lselfgemma/talk/domain/roleplay/usecase/ImportStChatJsonlIntoSessionUseCase;", "exportStChatJsonlFromSessionUseCase", "Lselfgemma/talk/domain/roleplay/usecase/ExportStChatJsonlFromSessionUseCase;", "<init>", "(Landroid/content/Context;Lselfgemma/talk/domain/roleplay/repository/ConversationRepository;Lselfgemma/talk/domain/roleplay/repository/RoleRepository;Lselfgemma/talk/domain/roleplay/usecase/EnsureRoleplaySeedDataUseCase;Lselfgemma/talk/domain/roleplay/usecase/ImportStChatJsonlIntoSessionUseCase;Lselfgemma/talk/domain/roleplay/usecase/ExportStChatJsonlFromSessionUseCase;)V", "feedbackState", "Lkotlinx/coroutines/flow/MutableStateFlow;", "Lselfgemma/talk/feature/roleplay/sessions/SessionsUiState;", "uiState", "Lkotlinx/coroutines/flow/StateFlow;", "getUiState", "()Lkotlinx/coroutines/flow/StateFlow;", "togglePin", "", "sessionId", "", "archiveSession", "deleteSession", "importChatJsonl", "uri", "exportChatJsonl", "appString", "resId", "", "args", "", "", "(I[Ljava/lang/Object;)Ljava/lang/String;", "Companion", "app_debug"})
@dagger.hilt.android.lifecycle.HiltViewModel()
public final class SessionsViewModel extends androidx.lifecycle.ViewModel {
    @org.jetbrains.annotations.NotNull()
    private final android.content.Context appContext = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.repository.ConversationRepository conversationRepository = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.ImportStChatJsonlIntoSessionUseCase importStChatJsonlIntoSessionUseCase = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.ExportStChatJsonlFromSessionUseCase exportStChatJsonlFromSessionUseCase = null;
    @org.jetbrains.annotations.NotNull()
    private static final java.lang.String TAG = "SessionsViewModel";
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.MutableStateFlow<selfgemma.talk.feature.roleplay.sessions.SessionsUiState> feedbackState = null;
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.StateFlow<selfgemma.talk.feature.roleplay.sessions.SessionsUiState> uiState = null;
    @org.jetbrains.annotations.NotNull()
    public static final selfgemma.talk.feature.roleplay.sessions.SessionsViewModel.Companion Companion = null;
    
    @javax.inject.Inject()
    public SessionsViewModel(@dagger.hilt.android.qualifiers.ApplicationContext()
    @org.jetbrains.annotations.NotNull()
    android.content.Context appContext, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.repository.ConversationRepository conversationRepository, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.repository.RoleRepository roleRepository, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.EnsureRoleplaySeedDataUseCase ensureRoleplaySeedData, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.ImportStChatJsonlIntoSessionUseCase importStChatJsonlIntoSessionUseCase, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.ExportStChatJsonlFromSessionUseCase exportStChatJsonlFromSessionUseCase) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    public final kotlinx.coroutines.flow.StateFlow<selfgemma.talk.feature.roleplay.sessions.SessionsUiState> getUiState() {
        return null;
    }
    
    public final void togglePin(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId) {
    }
    
    public final void archiveSession(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId) {
    }
    
    public final void deleteSession(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId) {
    }
    
    public final void importChatJsonl(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId, @org.jetbrains.annotations.NotNull()
    java.lang.String uri) {
    }
    
    public final void exportChatJsonl(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId, @org.jetbrains.annotations.NotNull()
    java.lang.String uri) {
    }
    
    private final java.lang.String appString(@androidx.annotation.StringRes()
    int resId, java.lang.Object... args) {
        return null;
    }
    
    @kotlin.Metadata(mv = {2, 2, 0}, k = 1, xi = 48, d1 = {"\u0000\u0012\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0003\n\u0002\u0010\u000e\n\u0000\b\u0086\u0003\u0018\u00002\u00020\u0001B\t\b\u0002\u00a2\u0006\u0004\b\u0002\u0010\u0003R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082T\u00a2\u0006\u0002\n\u0000\u00a8\u0006\u0006"}, d2 = {"Lselfgemma/talk/feature/roleplay/sessions/SessionsViewModel$Companion;", "", "<init>", "()V", "TAG", "", "app_debug"})
    public static final class Companion {
        
        private Companion() {
            super();
        }
    }
}

File: Android/src/app/src/main/java/selfgemma/talk/feature/roleplay/sessions/SessionsViewModel.kt
package selfgemma.talk.feature.roleplay.sessions

import android.content.Context
import android.util.Log
import androidx.annotation.StringRes
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import selfgemma.talk.R
import selfgemma.talk.domain.roleplay.repository.ConversationRepository
import selfgemma.talk.domain.roleplay.repository.RoleRepository
import selfgemma.talk.domain.roleplay.usecase.EnsureRoleplaySeedDataUseCase
import selfgemma.talk.domain.roleplay.usecase.ExportStChatJsonlFromSessionUseCase
import selfgemma.talk.domain.roleplay.usecase.ImportStChatJsonlIntoSessionUseCase
import selfgemma.talk.domain.roleplay.model.primaryAvatarUri

data class SessionListItemUiState(
  val id: String,
  val title: String,
  val roleName: String,
  val avatarUri: String?,
  val pinned: Boolean,
  val updatedAt: Long,
  val lastMessage: String,
)

data class SessionsUiState(
  val loading: Boolean = true,
  val sessions: List<SessionListItemUiState> = emptyList(),
  val statusMessage: String? = null,
  val errorMessage: String? = null,
)

@HiltViewModel
class SessionsViewModel
@Inject
constructor(
  @ApplicationContext private val appContext: Context,
  private val conversationRepository: ConversationRepository,
  roleRepository: RoleRepository,
  ensureRoleplaySeedData: EnsureRoleplaySeedDataUseCase,
  private val importStChatJsonlIntoSessionUseCase: ImportStChatJsonlIntoSessionUseCase,
  private val exportStChatJsonlFromSessionUseCase: ExportStChatJsonlFromSessionUseCase,
) : ViewModel() {
  companion object {
    private const val TAG = "SessionsViewModel"
  }

  private val feedbackState = MutableStateFlow(SessionsUiState(loading = false))

  val uiState: StateFlow<SessionsUiState> =
    combine(
      conversationRepository.observeSessions(),
      roleRepository.observeRoles(),
      feedbackState,
    ) { sessions, roles, feedback ->
      val rolesById = roles.associateBy { it.id }
      SessionsUiState(
        loading = false,
        sessions =
          sessions.map { session ->
            val role = rolesById[session.roleId]
            SessionListItemUiState(
              id = session.id,
              title = session.title,
              roleName = role?.name ?: appString(R.string.sessions_unknown_role),
              avatarUri = role?.primaryAvatarUri(),
              pinned = session.pinned,
              updatedAt = session.updatedAt,
              lastMessage =
                session.lastAssistantMessageExcerpt
                  ?: session.lastUserMessageExcerpt
                  ?: session.lastSummary
                  ?: appString(R.string.sessions_no_messages),
            )
          },
        statusMessage = feedback.statusMessage,
        errorMessage = feedback.errorMessage,
      )
    }
      .stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5_000),
        initialValue = SessionsUiState(),
      )

  init {
    viewModelScope.launch {
      ensureRoleplaySeedData()
    }
  }

  fun togglePin(sessionId: String) {
    viewModelScope.launch {
      val session = conversationRepository.getSession(sessionId) ?: return@launch
      val now = System.currentTimeMillis()
      val nextPinned = !session.pinned
      Log.d(TAG, "togglePin sessionId=$sessionId fromPinned=${session.pinned} toPinned=$nextPinned")
      conversationRepository.updateSession(
        session.copy(pinned = nextPinned, updatedAt = now, lastMessageAt = session.lastMessageAt)
      )
    }
  }

  fun archiveSession(sessionId: String) {
    viewModelScope.launch {
      conversationRepository.archiveSession(sessionId)
    }
  }

  fun deleteSession(sessionId: String) {
    viewModelScope.launch {
      conversationRepository.deleteSession(sessionId)
    }
  }

  fun importChatJsonl(sessionId: String, uri: String) {
    viewModelScope.launch {
      runCatching {
        importStChatJsonlIntoSessionUseCase.importIntoSession(sessionId = sessionId, uri = uri)
      }
        .onSuccess {
          feedbackState.update {
            it.copy(
              statusMessage = appString(R.string.sessions_status_imported),
              errorMessage = null,
            )
          }
        }
        .onFailure { error ->
          feedbackState.update {
            it.copy(
              statusMessage = null,
              errorMessage = error.message ?: appString(R.string.sessions_error_import_failed),
            )
          }
        }
    }
  }

  fun exportChatJsonl(sessionId: String, uri: String) {
    viewModelScope.launch {
      runCatching {
        exportStChatJsonlFromSessionUseCase.exportFromSession(sessionId = sessionId, uri = uri)
      }
        .onSuccess {
          feedbackState.update {
            it.copy(
              statusMessage = appString(R.string.sessions_status_exported),
              errorMessage = null,
            )
          }
        }
        .onFailure { error ->
          feedbackState.update {
            it.copy(
              statusMessage = null,
              errorMessage = error.message ?: appString(R.string.sessions_error_export_failed),
            )
          }
        }
    }
  }

  private fun appString(@StringRes resId: Int, vararg args: Any): String {
    return appContext.getString(resId, *args)
  }
}
